### ISSUE TITLE
Know a partial message is received on end of file?

### ISSUE THREAD

**Issue by vinniefalco** [vinniefalco/Beast#197]
During HTTP read if the end of file is reached in the middle of a message, should we return `error::eof` to the caller or should it be handled as a parsing error, and then return `error::eof` on the next call to read?

**Comment by djarek** [comment]
Receiving an eof does not necessarily mean the connection entered an "invalid" state. A peer (this mostly true for clients) can call ```shutdown()``` on their write end of the socket as a way to indicate end of message to the other peer (as an alternative to providing Content-length, this is mostly used for backwards compatibility with HTTP/1.0 or dumb implementations that say they are HTTP/1.1 compliant but actually aren't). So this input:
```
POST /do-stuff HTTP/1.1
Host: myhost.com
Content-type: application/json

{"hello" : "world"}
[EOF]
```
is a "valid" HTTP message. Of course, the same is true if the request body were empty. I think the only situation where you should treat EOF as an error that has to be returned after parsing is if we encounter:

1. EOF before final header line.
2. EOF before ```Content-length``` bytes are transferred (assuming Content-Length is present in the header).

This might be obvious, but another "valid" message is this:
```
POST /do-stuff HTTP/1.1
Host: myhost.com
Content-type: application/json
Content-length: 2

{}
[EOF]
```

There's a reason why I put the word "valid" in quotes. They're not really valid if we want to be pedantic about RFC 7230's [message length](https://tools.ietf.org/html/rfc7230#section-3.3.3):

> A server MAY reject a request that contains a message body but not a
   Content-Length by responding with 411 (Length Required).

The emphasis on the word MAY means that a general-purpose library like Beast should (IMO) consider such messages to be valid.

**Comment by 2underscores-vic** [comment]
Some info about using half-close for request-messages in practice:

http://stackoverflow.com/questions/14069379/does-http-1-1-allow-delimiting-end-of-request-by-half-closing-the-connection
http://mailman.nginx.org/pipermail/nginx/2008-September/007358.html

tl;dr: seems like it's a weird scenario

**Comment by vinniefalco** [comment]
Its going to be a parse error, returning `error::partial_message`. Which implies **eof**.
